database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -41,8 +42,11 @@ public FirebaseIndexArray(Query keyRef, Query dataRef) {
     @Override
     public void cleanup() {
         super.cleanup();
-        for (Map.Entry<Query, ValueEventListener> entry : mRefs.entrySet()) {
+        final Iterator<Map.Entry<Query, ValueEventListener>> it = mRefs.entrySet().iterator();
+        while (it.hasNext()) {
+            final Map.Entry<Query, ValueEventListener> entry = it.next();
             entry.getKey().removeEventListener(entry.getValue());
+            it.remove();
```

Comment:
Does calling `Iterator#remove()` on an `Iterator` created from `.entrySet()` actually delete from the underlying map?  Maybe we should add a test that after `cleanup()` we verify `mRefs` is empty.